Success story: 84codes (#788)
crystal-lang/crystal-website: _posts/success_stories/2024-05-20-lavinmq-understanding-code.md

---
title: "LavinMQ: Understanding code to get the best performance"

> summary: "Building LavinMQ in Crystal was an active choice, as maintaining control over the codebase was a high priority."

Reviewer: **thought:** To me this does not feels very self-concious, it has a bit a tune of an apology maybe.

How about something more catchy?
```suggestion
summary: "Crystal was chosen for maintaining control over the entire codebase."
```

author: beta-ziliani
image: /assets/blog/2024-05-20-lavinmq-terminal.png
company: 84codes
categories: success
---

> **Building LavinMQ in Crystal was an active choice, as maintaining control over the codebase was a high priority. A decade of learning has prioritized fast updates and easy bug fixes. At the same time, choosing Crystal provided the opportunity to enhance application performance.**

Reviewer: **issue:** I'm not sure highlighting the lead paragraph is a good idea.
We can introduce this, but it should be a concious feature in the visual framework, not an ad-hoc implementation for this single post.

Author: I think this summary paragraph makes sense, and it must be highlighted somehow. If any, I would rather keep it as is, and later modify the fwk to include thiis feature.

Reviewer: Why "must"? It works perfectly fine without the bold face.
![grafik](https://github.com/crystal-lang/crystal-website/assets/466378/0023f296-edce-4eb0-a418-24c18a802981)

Author: there's a history break between the paragraphs. the bold was marking the paragraph as something different than an intro. but it's a nitpick really. I removed the boldface.

> Messages are written to the disk really fast and the caching is managed by the OS itself. This leads to a performance boost while keeping RAM footprint low.

But the other half, as mentioned in [their blog](https://lavinmq.com/blog/crystal-clear-message-brokering), is due to it being developed in Crystal. And this comes in several interesting ways.

1. Crystal is efficiently compiled thanks to its LLVM backend. This enables aggressive optimizations, like inlining methods and removing indirections.

2. Crystal helps developers allocate their data either on the heap or on the stack. Heap allocations are expensive, so allocating data on the stack in the hot-path of the program brings an important boost in performance. Crystal makes it easy to choose where data should be stored.

3. Crystal's stdlib and compiler are written in Crystal itself. One of Crystal's golden characteristics is its readability, making it possible to read the code that is going to be run by the program without having to resort to another, typically lower-level  language.

> The latter is the main point of this post. In fact, the documentation of the standard library points to the code, making the performance implications of calling a certain method explicit. As an example that will come up later, the documentation for the `file` method of the `Digest` class can be found [here](https://crystal-lang.org/api/1.12.1/Digest.html#file%28file_name%3APath%7CString%29%3Aself-instance-method). Clicking on [[View source](https://github.com/crystal-lang/crystal/blob/4cea10199/src/digest/digest.cr#L214)] we land in the method’s code.

Reviewer: **issue:** *here* links are bad. Link texts should be descriptive, and this is easy to achieve here.
I also took the liberty to shorten the label to the formal method reference instead of a verbal description for conciseness.

"As an example that will come up later," doesn't make sense here. We're directly using that example right now.

```suggestion
The latter is the main point of this post. In fact, the documentation of the standard library points to the code, making the performance implications of calling a certain method explicit. For example, look at the [documentation for `Digest.file`](https://crystal-lang.org/api/1.12.1/Digest.html#file%28file_name%3APath%7CString%29%3Aself-instance-method). Clicking on [*View source*](https://github.com/crystal-lang/crystal/blob/4cea10199/src/digest/digest.cr#L214) we land in the method’s code.

```

> The importance of being able to read, understand, and even contribute to the language and its stdlib is exemplified in [one](https://github.com/crystal-lang/crystal/pull/13780) of the [several PRs](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+author%3Acarlhoerberg+) that 84codes contributed to Crystal. Details are not important, but if you’re curious, they noted that in the aforementioned `Digest` class’s method there was a duplication of buffering. The following picture shows the contribution: you can find in green the added line and comment explaining why it makes sense to remove the buffering in the `io` object.

Author: ```suggestion
The importance of being able to read, understand, and even contribute to the language and its stdlib is exemplified in [one](https://github.com/crystal-lang/crystal/pull/13780) of the [several PRs](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+author%3Acarlhoerberg+) that 84codes contributed to Crystal. Details are not important, but if you’re curious, they noted that in the aforementioned `Digest` class’s method there was a duplication of buffering. The following extract shows the main part of the contribution: you can find in green the added line and comment explaining why it makes sense to remove the buffering in the `io` object.

```

> ![diff](/assets/blog/2024-05-20-digest-diff.png)

Reviewer: **suggestion:** Display the diff directly. That's more accessible and readable than a picture.

````suggestion
```diff
diff --git a/src/digest/digest.cr b/src/digest/digest.cr
index e6a401e90545..bc38599da7f8 100644
--- a/src/digest/digest.cr
+++ b/src/digest/digest.cr
@@ -213,6 +213,7 @@ abstract class Digest
   # Reads the file's content and updates the digest with it.
   def file(file_name : Path | String) : self
     File.open(file_name) do |io|
+      # `#update` works with big buffers so there's no need for additional read buffering in the file
+      io.read_buffering = false
       self << io
     end
   end
```
````

This needs updates to the reference above (`s/picture/diff/`) and remove of the image file.

One little line, yet it can have a noticeable impact on the application. Having the same language for the application and the libraries the code depends upon is a great advantage in understanding the behavior of the code and its performance implications.

Before closing, we’d like to note that 84codes is not only responsible for numerous improvements to the standard library: 84codes is also the main sponsor supporting the development of Crystal. In order to push the limits of the language even further, at the time of writing they are supporting a large effort to improve [the parallel execution](https://crystal-lang.org/2024/02/09/84codes-manas-mt/) of Crystal programs. Perhaps the next version of LavinMQ will push the limits even further with the help of the Crystal Team!

> {% include components/testimonial-profile.html handle="carl" role="CEO, 84codes" %}

Author: I forgot to check in this file.
